### ISSUE TITLE
how to access the CRS of the writer?

### ISSUE THREAD

**Issue by hugoledoux** [safesoftware/fme-CityJSON#35]
@KevinWiebe I don't know how to access the CRS of the writer that was setup by the user in the writer parameters, see below.

It's surely one variable in the gMappingFile:
```
gMappingFile->fetchWithPrefix(writerKeyword_.c_str(), writerTypeName_.c_str(), WHICH_ONE, *pv);
```
but I don't know which one.

![2020-01-21 at 16 02](https://user-images.githubusercontent.com/1546518/72815984-cacdbb80-3c67-11ea-856d-22e6704df91f.jpg)

**Comment by KevinWiebe** [comment]
First, I assume you already know how to get the parameters like compression and Important Digits.

Something like this in the writer's "open" method:

```   IFMEString* compress = gFMESession->createString();
   // provide a default compression if none can be found 
   if (!gMappingFile->fetchWithPrefix(writerKeyword_.c_str(), writerTypeName_.c_str(), kDstCompressionParamTag, *compress))
   {
      compress->set("No", 2);
   }

   gFMESession->destroyString(compress); compress = nullptr;
```


With this in fmecityjsonpriv.h

```const static char* const kDstCompressionParamTag = "_USE_COMPRESSION";
const static char* const kDstImportantDigitsParamTag = "_IMPORTANT_DIGITS";
```

**Comment by KevinWiebe** [comment]
But to your point, the `Coordinate System` parameter is special.  It is not fetched like any of the others.

This is more of a global setting on ALL writers, to a central mechanism.  It directs the FME core on how to actually reproject (or "tag" features) with a coordinate system before they are passed in to the writer.

So if you set this to EPSG-7415, it means all features will be reprojected to EPSG-7415 before they get sent to your `FMECityJSONWriter::write()` method.

If you want to know what this was set to, you have to wait until you get the first feature sent to you, and then ask that feature what coordinate system it has.

Does this make sense?

If you need this to write out to a header (like we do sometimes), we have a common pattern we use:
- In the "open" method, set up all you need for the header, except the things you need to wait for the first feature for.  But don't write out anything at all yet.
- In the "write" call, check a boolean "haveWrittenHeader" or something.  If it is false, then scrape all you need and write out the header before processing the feature itself.  (And set the flag "haveWrittenHeader" to true, so you don't do it again for the next feature in.)

Let me know if you need any more info, but I think this should get you on your way here.

**Comment by hugoledoux** [comment]
The other parameters we have already coded for reader/writer, works like a charm.

For CRS, I guess I was over-thinking it to be honest. I thought it was to say also (besides reprojecting) that "hey this dataset for which you don't know the CRS, it's actually EPSG:7415!". One can do that with QGIS, say because the CRS is not in the metadata of a file, or one doesn't have the .prj file for their .shp

Can that happen with FME? I guess yes! Then the first feature you get has no CRS, and you can't use that.

**Comment by KevinWiebe** [comment]
I think FME works with CRS a bit differently (and is at least _trying_ to be a bit more helpful).

If you set the CRS in the writers settings box, then it will two two things:
- all features that are sent to the writer with NO CRS will not be changed, but will be tagged with that CRS.
- all features that are sent to the writer with their own CRS will be reprojected to the one on the writer's setting.

In either case, every single feature going into the writer should have the same CRS.
So you are right, all you need to do is look at the first one.


However, if the writer setting of CRS is not set to anything, then I believe all features are sent into the writer with whatever they've got.  Some may have one set, some may not, and they may not all agree.

In this case, one reasonable thing you could do is to just check the first feature's CRS and start writing out - and if you come across one with no CRS or a different one, give an error message and quit.  Or if you are more friendly, you can tag or reproject them yourself in the writer.  Up to you.

I hope this gives you enough information to move forward here.

**Comment by KevinWiebe** [comment]
I believe I fixed this last week with pull request #64
